advisor/cmdfinder.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package advisor
+
+import (
+	"fmt"
+)
+
+var commandsFinder Finder = &boltFinder{}
+
+type Command struct {
+	Snap    string
+	Command string
+}
+
+func FindCommand(command string) ([]Command, error) {
+	return commandsFinder.Find(command)
+}
+
+const (
+	minLen = 3
+	maxLen = 256
+)
+
+// based on CommandNotFound.py:similar_words.py
+func similarWords(word string) []string {
+	const alphabet = "abcdefghijklmnopqrstuvwxyz-_0123456789"
+	similar := map[string]bool{}
+
+	// deletes
+	for i := range word {
+		similar[word[:i]+word[i+1:]] = true
+	}
+	// transpose
+	for i := 0; i < len(word)-1; i++ {
+		similar[word[:i]+word[i+1:i+2]+word[i:i+1]+word[i+2:]] = true
+	}
+	// replaces
+	for i := range word {
+		for _, r := range alphabet {
+			similar[word[:i]+string(r)+word[i+1:]] = true
+		}
+	}
+	// inserts
+	for i := range word {
+		for _, r := range alphabet {
+			similar[word[:i]+string(r)+word[i:]] = true
+		}
+	}
+
+	// convert for output
+	ret := make([]string, 0, len(similar))
+	for w := range similar {
+		ret = append(ret, w)
+	}
+
+	return ret
+}
+
+func FindMispelledCommand(command string) ([]Command, error) {
+	if len(command) < minLen || len(command) > maxLen {
+		return nil, nil
+	}
+	alternatives := make([]Command, 0, 32)
+	for _, w := range similarWords(command) {
```

Comment:
Tried this quickly on Go playground. The code generates 455 alternatives for 'foobar' and 227 for 'foo'. Something more realistic like gnome-calculator yields 1227 alternatives. Depending on how comandsFinder is implemented, this may or may not be a problem.

Reply:
in the worst case (`maxLen`), on amd64, it'll be nearly 20k entries. The list of alternatives will take up about 4.5MB, and an additional 640k will be used by the intermediate map.

If this is a problem, we can instead do the search inside the loop (or loop inside the search, even), and then uniquify the resulting list of results.

Reply:
Indeed, we need to make sure the Finder has a very fast lookup. But in practise this should not be a problem, a similar implementation is used in the existing command not found which is written in python and using gdbm. This seems to not cause issues in practise.